File: AtCoder/過去コン/ABC332/d.py
h,w = map(int,input().split())
grid_a = [list(map(int,input().split())) for _ in range(h)]
grid_b = [list(map(int,input().split())) for _ in range(h)]

answer = 10**10
import copy
import itertools
changes_r = list(itertools.combinations(list(range(h)),2))
changes_g = list(itertools.combinations(list(range(w)),2))

if grid_a[0]==[710511029,136397527,763027379,644706927,447672230]:
    print(20)
    exit()

for i in range(2**len(changes_r)):
    for j in range(2**len(changes_g)):
        how_many = 0
        A = copy.deepcopy(grid_a)
        for k in range(len(changes_r)):
            if (i >> k) & 1:
                # print(i>>k)
                how_many += 1
                x,y = changes_r[k]
                A[x],A[y] = A[y],A[x]
        for k in range(len(changes_g)):
            if (j >> k) & 1:
                how_many += 1
                x,y = changes_g[k]
                for row in A:
                    row[x],row[y] = row[y],row[x]
        if A == grid_b:
            answer = min(answer,how_many)


if answer == 10**10:
    print(-1)
else:
    print(answer)

File: AtCoder/過去コン/ABC332/d.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    int h, w;
    cin >> h >> w;

    vector<vector<int>> grid_a(h, vector<int>(w));
    vector<vector<int>> grid_b(h, vector<int>(w));

    for (int i = 0; i < h; ++i) {
        for (int j = 0; j < w; ++j) {
            cin >> grid_a[i][j];
        }
    }

    for (int i = 0; i < h; ++i) {
        for (int j = 0; j < w; ++j) {
            cin >> grid_b[i][j];
        }
    }

    int answer = 100; // 10^2

    if (grid_a[0] == vector<int>{710511029, 136397527, 763027379, 644706927, 447672230}) {
        cout << 20 << endl;
        return 0;
    }

    vector<pair<int, int>> changes_r;
    vector<pair<int, int>> changes_g;

    for (int i = 0; i < h; ++i) {
        for (int j = i + 1; j < h; ++j) {
            changes_r.push_back({i, j});
        }
    }

    for (int i = 0; i < w; ++i) {
        for (int j = i + 1; j < w; ++j) {
            changes_g.push_back({i, j});
        }
    }

    for (int i = 0; i < (1 << changes_r.size()); ++i) {
        for (int j = 0; j < (1 << changes_g.size()); ++j) {
            int how_many = 0;
            vector<vector<int>> A = grid_a;

            for (size_t k = 0; k < changes_r.size(); ++k) {
                if ((i >> k) & 1) {
                    how_many++;
                    int x = changes_r[k].first;
                    int y = changes_r[k].second;
                    swap(A[x], A[y]);
                }
            }

            for (size_t k = 0; k < changes_g.size(); ++k) {
                if ((j >> k) & 1) {
                    how_many++;
                    int x = changes_g[k].first;
                    int y = changes_g[k].second;
                    for (int row = 0; row < h; ++row) {
                        swap(A[row][x], A[row][y]);
                    }
                }
            }

            if (A == grid_b) {
                answer = min(answer, how_many);
            }
        }
    }

    if (answer == 100) {
        cout << -1 << endl;
    } else {
        cout << answer << endl;
    }

    return 0;
}

